oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/writer/IndexWriterPool.java

```diff
@@ -0,0 +1,407 @@
+            this.operations = operations;
+        }
+    }
+
+    private static abstract class Operation {
+        final LuceneIndexWriter delegate;
+
+        public Operation(LuceneIndexWriter delegate) {
+            this.delegate = delegate;
+        }
+
+        abstract void execute() throws IOException;
+    }
+
+    private static class UpdateOperation extends Operation {
+        private final String path;
+        private final Iterable<? extends IndexableField> doc;
+
+        UpdateOperation(LuceneIndexWriter delegate, String path, Iterable<? extends IndexableField> doc) {
+            super(delegate);
+            this.path = path;
+            this.doc = doc;
+        }
+
+        @Override
+        public void execute() throws IOException {
+            delegate.updateDocument(path, doc);
+        }
+    }
+
+    private static class DeleteOperation extends Operation {
+        private final String path;
+
+        DeleteOperation(LuceneIndexWriter delegate, String path) {
+            super(delegate);
+            this.path = path;
+        }
+
+        @Override
+        public void execute() throws IOException {
+            delegate.deleteDocuments(path);
+        }
+    }
+
+    private static class CloseResult {
+        // Either result or error is non-null. The two constructors enforce this invariant.
+        final Boolean result;
+        final Throwable error;
+
+        CloseResult(boolean result) {
+            this.result = result;
+            this.error = null;
+        }
+
+        CloseResult(Throwable error) {
+            this.result = null;
+            this.error = error;
+        }
+
+        @Override
+        public String toString() {
+            return "CloseResult{" +
+                    "result=" + result +
+                    ", error=" + error +
+                    '}';
+        }
+    }
+
+    private static class CloseWriterOperation extends Operation {
+        private final long timestamp;
+        private final SynchronousQueue<CloseResult> sync;
+
+        /**
+         * The close operation should be synchronous and applied only after all the write operations for this writer
+         * are processed.
+         *
+         * @param sync A synchronous queue used to wait for the result of the close operation.
+         */
+        CloseWriterOperation(LuceneIndexWriter delegate, long timestamp, SynchronousQueue<CloseResult> sync) {
+            super(delegate);
+            this.timestamp = timestamp;
+            this.sync = sync;
+        }
+
+        @Override
+        public void execute() {
+            try {
+                try {
+                    boolean closeResult = delegate.close(timestamp);
+                    sync.put(new CloseResult(closeResult));
+                } catch (IOException e) {
+                    sync.put(new CloseResult(e));
+                }
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
```

Comment:
I wonder if this is the right thing to do... Why not 
```suggestion
                LOG.warn("[{}] Interrupted while closing", timestamp, e);
                sync.put(new CloseResult(new IOException(e)));
```
or something like that?

Reply:
`sync.put()` throws an `InterrupedException`, so we cannot use it on the outermost exception handler. If `sync.put()` fails, then I don't think we can do anything else other than quitting with a `RuntimeException`.